File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/camera/CameraActivity.java
package kr.co.yogiyo.rookiephotoapp.camera;

import android.content.Intent;

import android.graphics.Bitmap;

import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.otaliastudios.cameraview.AspectRatio;
import com.otaliastudios.cameraview.CameraListener;
import com.otaliastudios.cameraview.CameraUtils;
import com.otaliastudios.cameraview.CameraView;
import com.otaliastudios.cameraview.Facing;
import com.otaliastudios.cameraview.Flash;
import com.otaliastudios.cameraview.Gesture;
import com.otaliastudios.cameraview.GestureAction;
import com.otaliastudios.cameraview.SizeSelector;
import com.otaliastudios.cameraview.SizeSelectors;

import java.util.Arrays;
import java.util.List;

import kr.co.yogiyo.rookiephotoapp.R;
import kr.co.yogiyo.rookiephotoapp.camera.capture.PreviewActivity;
import kr.co.yogiyo.rookiephotoapp.camera.capture.ResultHolder;

public class CameraActivity extends AppCompatActivity implements View.OnClickListener {

    private static final List<Integer> DELAY_DURATIONS = Arrays.asList(0, 2, 5, 10);
    private static final int DEFAULT_DELAY = 0;

    private Handler timerHandler;

    private int captureTimer;
    private int currentCaptureID;
    private int captureDelay;
    private int flashType;

    private CameraView cameraView;
    private FrameLayout controlButtonsFrameLayout;
    private Button backButton;
    private Button flashButton;
    private Button timerButton;
    private Button captureButton;
    private Button changeCameraButton;
    private TextView timerMessageTextView;
    private FrameLayout darkScreenFrame;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_camera);

        initView();

        initialize();
    }

    private void initView() {
        cameraView = findViewById(R.id.camera);
        controlButtonsFrameLayout = findViewById(R.id.frame_control_buttons);
        backButton = findViewById(R.id.btn_back);
        flashButton = findViewById(R.id.btn_flash);
        timerButton = findViewById(R.id.btn_timer);
        captureButton = findViewById(R.id.btn_capture);
        changeCameraButton = findViewById(R.id.btn_change_camera);
        timerMessageTextView = findViewById(R.id.text_timer_message);
        darkScreenFrame = findViewById(R.id.frame_dark_screen);

        backButton.setOnClickListener(this);
        flashButton.setOnClickListener(this);
        timerButton.setOnClickListener(this);
        captureButton.setOnClickListener(this);
        changeCameraButton.setOnClickListener(this);

        initCameraView();

        updateDelayButton();
    }

    private void initCameraView() {
        cameraView.addCameraListener(new CameraListener() {
            @Override
            public void onPictureTaken(byte[] jpeg) {
                if (jpeg != null) {
                    CameraUtils.decodeBitmap(jpeg, new CameraUtils.BitmapCallback() {
                        @Override
                        public void onBitmapReady(Bitmap bitmap) {
                            if (bitmap == null) {
                                finish();
                                return;
                            }
                            ResultHolder.setBitmap(bitmap);
                            Intent intent = new Intent(CameraActivity.this, PreviewActivity.class);
                            startActivity(intent);
                        }
                    });
                }
            }
        });
        cameraView.mapGesture(Gesture.TAP, GestureAction.FOCUS_WITH_MARKER);

        SizeSelector ratio = SizeSelectors.aspectRatio(AspectRatio.of(3, 4), 0);
        SizeSelector result = SizeSelectors.or(ratio, SizeSelectors.biggest());
        cameraView.setPictureSize(result);
    }

    private void initialize() {
        timerHandler = new Handler();
        captureTimer = 0;
        currentCaptureID = 0;
        captureDelay = DEFAULT_DELAY;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_back:
                onBackPressed();
                break;
            case R.id.btn_flash:
                setFlashNext();
                break;
            case R.id.btn_timer:
                setCaptureDelayNext();
                break;
            case R.id.btn_capture:
                capture();
                break;
            case R.id.btn_change_camera:
                changeFacing();
                break;
        }
    }

    @Override
    public void onBackPressed() {
        if (timerHandler.hasMessages(0)) {
            finishDelayCapture();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        cameraView.start();
    }

    @Override
    protected void onPause() {
        super.onPause();
        cameraView.stop();
        if (timerHandler.hasMessages(0)) {
            finishDelayCapture();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cameraView.destroy();
    }

    private void captureNow() {
        cameraView.capturePicture();
        startBlinkAnimation();
    }

    private void capture() {
        if (captureDelay == DEFAULT_DELAY) {
            captureNow();
        } else {
            captureAfterDelay(captureDelay);
        }
    }

    private void setCaptureDelayNext() {
        int index = DELAY_DURATIONS.indexOf(captureDelay);
        if (index < 0) {
            captureDelay = DEFAULT_DELAY;
        } else {
            captureDelay = DELAY_DURATIONS.get((index + 1) % DELAY_DURATIONS.size());
        }
        updateDelayButton();
    }

    private void updateDelayButton() {
        if (captureDelay == 0) {
            timerButton.setText(getString(R.string.text_timer));
        } else {
            String timerButtonTextFormat = getString(R.string.text_timer_button_text_format);
            timerButton.setText(String.format(timerButtonTextFormat, captureDelay));
        }
    }

    private void captureAfterDelay(int delay) {
        timerMessageTextView.setVisibility(View.VISIBLE);
        captureTimer = delay;
        updateTimerMessage();
        currentCaptureID++;
        timerHandler.postDelayed(makeDecrementTimerFunction(currentCaptureID), 1000);

        setControlButtonsVisibility(false);
    }

    private Runnable makeDecrementTimerFunction(final int captureID) {
        return new Runnable() {
            public void run() {
                decrementTimer(captureID);
            }
        };
    }

    private void decrementTimer(final int captureID) {
        if (captureID != currentCaptureID) {
            return;
        }
        --captureTimer;
        if (captureTimer == 0) {
            captureNow();
            finishDelayCapture();
        } else if (captureTimer > 0) {
            updateTimerMessage();
            timerHandler.postDelayed(makeDecrementTimerFunction(captureID), 1000);
        }
    }

    private void updateTimerMessage() {
        String timerMessageFormat = getString(R.string.text_timer_message_format);
        timerMessageTextView.setText(String.format(timerMessageFormat, captureTimer));
    }

    private void setControlButtonsVisibility(boolean allowSave) {
        controlButtonsFrameLayout.setVisibility(allowSave ? View.VISIBLE : View.GONE);
    }

    private void finishDelayCapture() {
        timerHandler.removeCallbacksAndMessages(null);
        timerMessageTextView.setVisibility(View.GONE);
        setControlButtonsVisibility(true);
    }

    private void setFlashNext() {
        flashType = (flashType + 1) % Flash.values().length;
        switch (flashType) {
            case 0:
                cameraView.setFlash(Flash.OFF);
                updateFlashButton("");
                break;
            case 1:
                cameraView.setFlash(Flash.ON);
                updateFlashButton(Flash.ON.name());
                break;
            case 2:
                cameraView.setFlash(Flash.AUTO);
                updateFlashButton(Flash.AUTO.name());
                break;
            case 3:
                cameraView.setFlash(Flash.TORCH);
                updateFlashButton(Flash.TORCH.name());
                break;
            default:
                break;
        }
    }

    private void updateFlashButton(String flashTypeString) {
        String flashButtonTextFormat = getString(R.string.text_flash_button_text_format);
        flashButton.setText(String.format(flashButtonTextFormat, flashTypeString));
    }

    private void changeFacing() {
        if (cameraView.getFacing() == Facing.FRONT) {
            cameraView.setFacing(Facing.BACK);
            updateFacingButton(Facing.BACK.name());
        } else {
            cameraView.setFacing(Facing.FRONT);
            updateFacingButton(Facing.FRONT.name());
        }
    }

    private void updateFacingButton(String facingTypeString) {
        String facingButtonTextFormat = getString(R.string.text_facing_button_text_format);
        changeCameraButton.setText(String.format(facingButtonTextFormat, facingTypeString));
    }

    private void startBlinkAnimation() {
        darkScreenFrame.setVisibility(View.VISIBLE);

        Animation blinkAnimation = AnimationUtils.loadAnimation(CameraActivity.this, R.anim.blink);
        blinkAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                darkScreenFrame.setVisibility(View.GONE);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        darkScreenFrame.startAnimation(blinkAnimation);
    }
}

File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/camera/CameraActivity.kt
package kr.co.yogiyo.rookiephotoapp.camera

import android.app.Activity
import android.content.Intent
import android.databinding.DataBindingUtil
import android.graphics.PointF
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.otaliastudios.cameraview.Flash
import com.otaliastudios.cameraview.Grid
import com.otaliastudios.cameraview.CameraListener
import com.otaliastudios.cameraview.CameraUtils
import com.otaliastudios.cameraview.SizeSelectors
import com.otaliastudios.cameraview.AspectRatio
import kotlinx.android.synthetic.main.activity_camera.*
import kr.co.yogiyo.rookiephotoapp.BaseActivity
import kr.co.yogiyo.rookiephotoapp.Constants
import kr.co.yogiyo.rookiephotoapp.GlobalApplication
import kr.co.yogiyo.rookiephotoapp.R
import kr.co.yogiyo.rookiephotoapp.camera.capture.PreviewActivity
import kr.co.yogiyo.rookiephotoapp.databinding.ActivityCameraBinding
import kr.co.yogiyo.rookiephotoapp.diary.DiaryEditActivity
import kr.co.yogiyo.rookiephotoapp.diary.main.DiariesActivity
import kr.co.yogiyo.rookiephotoapp.gallery.GalleryActivity
import kr.co.yogiyo.rookiephotoapp.gallery.GalleryFragment
import java.util.Calendar

class CameraActivity : BaseActivity() {

    private var backPressedStartTime = 0L

    private val cameraViewModel: CameraViewModel by lazy {
        CameraViewModel(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        (DataBindingUtil.setContentView(this, R.layout.activity_camera) as ActivityCameraBinding)
                .viewModel = cameraViewModel

        initView()

        initCameraView()

        cameraViewModel.initViewModel()
    }

    override fun onResume() {
        super.onResume()
        cameraViewModel.run {
            initButtonVisibility()
            updateGalleryButton()
        }

        camera.start()
    }

    override fun onPause() {
        super.onPause()
        camera.stop()
    }

    override fun onDestroy() {
        super.onDestroy()
        camera.destroy()
    }

    override fun onBackPressed() {
        when {
            cameraViewModel.timerCancel() -> return
            btn_show_more.hasFocus() -> btn_show_more.clearFocus()
            btn_capture_size.hasFocus() -> btn_capture_size.clearFocus()
            else -> {
                backPressedStartTime = Calendar.getInstance().timeInMillis.also {
                    if (it - backPressedStartTime < 2000 ||
                            GlobalApplication.globalApplicationContext.isFromDiary) {
                        super.onBackPressed()
                    } else {
                        showSnackbar(relative_root, getString(R.string.text_finish_snackbar))
                    }
                }
            }
        }
    }

    private fun initView() {

        btn_go_diary.setOnClickListener {
            val intent = Intent(this, DiariesActivity::class.java)
            startActivity(intent)
        }

        relative_flash.setOnClickListener {

            when (cameraViewModel.getNextFlashType()) {
                0 -> {
                    camera.flash = Flash.OFF
                    cameraViewModel.updateFlashButton(getString(R.string.text_flash), R.drawable.baseline_flash_off_white_24)
                }
                1 -> {
                    camera.flash = Flash.ON
                    cameraViewModel.updateFlashButton(getString(R.string.text_flash_ON), R.drawable.baseline_flash_on_white_24)
                }
                2 -> {
                    camera.flash = Flash.AUTO
                    cameraViewModel.updateFlashButton(getString(R.string.text_flash_AUTO), R.drawable.baseline_flash_auto_white_24)
                }
                3 -> {
                    camera.flash = Flash.TORCH
                    cameraViewModel.updateFlashButton(getString(R.string.text_flash_TORCH), R.drawable.baseline_flash_on_white_24)
                }
            }
        }

        relative_go_gallery.run {
            setOnClickListener {
                val intent = Intent(this@CameraActivity, GalleryActivity::class.java)
                startActivity(intent)
            }
        }

        btn_one_to_one.setOnClickListener {
            cameraViewModel.updateViewByCaptureSize(false)
            setCaptureSize(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, 1, 1)
        }

        btn_three_to_four.setOnClickListener {
            cameraViewModel.updateViewByCaptureSize(false)
            setCaptureSize(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, 3, 4)
        }

        btn_nine_to_sixteen.setOnClickListener {
            cameraViewModel.updateViewByCaptureSize(true)
            setCaptureSize(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 9, 16)
        }

        relative_grid.setOnClickListener {
            camera.run {
                grid = when (cameraViewModel.getNextGridType()) {
                    0 -> {
                        cameraViewModel.updateGridButton(context.getString(R.string.text_grid),
                                if (cameraViewModel.isCaptureSizeFull()) {
                                    R.drawable.baseline_grid_off_white_36
                                } else {
                                    R.drawable.baseline_grid_off_black_36
                                })
                        Grid.OFF
                    }
                    1 -> {
                        cameraViewModel.updateGridButton(context.getString(R.string.text_grid_three_by_three),
                                if (cameraViewModel.isCaptureSizeFull()) {
                                    R.drawable.baseline_grid_on_white_36
                                } else {
                                    R.drawable.baseline_grid_on_black_36
                                })
                        Grid.DRAW_3X3
                    }
                    2 -> {
                        cameraViewModel.updateGridButton(context.getString(R.string.text_grid_four_by_four),
                                if (cameraViewModel.isCaptureSizeFull()) {
                                    R.drawable.baseline_grid_on_white_36
                                } else {
                                    R.drawable.baseline_grid_on_black_36
                                })
                        Grid.DRAW_4X4
                    }
                    else -> {
                        cameraViewModel.updateGridButton(context.getString(R.string.text_grid_phi),
                                if (cameraViewModel.isCaptureSizeFull()) {
                                    R.drawable.baseline_grid_on_white_36
                                } else {
                                    R.drawable.baseline_grid_on_black_36
                                })
                        Grid.DRAW_PHI
                    }
                }
            }
        }

        btn_show_more.run {
            setOnClickListener {
                cameraViewModel.updateShowMoreLayout(if (relative_more_control_buttons.visibility == View.VISIBLE) {
                    clearFocus()
                    View.GONE
                } else {
                    requestFocus()
                    View.VISIBLE
                })
            }

            setOnFocusChangeListener { _, hasFocus ->
                cameraViewModel.updateShowMoreLayout(if (!hasFocus) View.GONE else View.VISIBLE)
            }
        }

        btn_capture_size.run {
            setOnClickListener {
                cameraViewModel.updateShowCaptureSizeLayout(if (relative_capture_size_buttons.visibility == View.VISIBLE) {
                    clearFocus()
                    View.GONE
                } else {
                    requestFocus()
                    View.VISIBLE
                })
            }

            setOnFocusChangeListener { _, hasFocus ->
                cameraViewModel.updateShowCaptureSizeLayout(if (!hasFocus) View.GONE else View.VISIBLE) }
        }

        relative_root.run {
            requestFocus()

            setOnClickListener {
                when {
                    btn_show_more.hasFocus() -> btn_show_more.clearFocus()
                    btn_capture_size.hasFocus() -> btn_capture_size.clearFocus()
                }
            }
        }
    }

    private fun initCameraView() {
        camera.run {
            addCameraListener(object : CameraListener() {
                override fun onPictureTaken(jpeg: ByteArray?) {
                    if (jpeg != null) {
                        startBlinkAnimation()
                        CameraUtils.decodeBitmap(jpeg, CameraUtils.BitmapCallback { bitmap ->
                            if (bitmap == null) {
                                finish()
                                return@BitmapCallback
                            }
                            val intent = Intent(this@CameraActivity, PreviewActivity::class.java)
                            PreviewActivity.capturedImageBitmap = bitmap
                            if (GlobalApplication.globalApplicationContext.isFromDiary) {
                                startActivityForResult(intent, Constants.REQUEST_DIARY_CAPTURE_PHOTO)
                            } else {
                                startActivity(intent)
                            }
                        })
                    }
                }

                override fun onFocusStart(point: PointF?) {
                    when {
                        btn_show_more.hasFocus() -> btn_show_more.clearFocus()
                        btn_capture_size.hasFocus() -> btn_capture_size.clearFocus()
                        else -> super.onFocusStart(point)
                    }
                }
            })
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        data?.let {
            when (resultCode) {
                Activity.RESULT_OK -> {
                    val intent = Intent(this@CameraActivity, DiaryEditActivity::class.java).apply {
                        this.data = it.data
                    }
                    setResult(Activity.RESULT_OK, intent)
                    finish()
                }
                Constants.RESULT_CAPTURED_PHOTO -> {
                    val intent = Intent(this@CameraActivity, DiaryEditActivity::class.java)
                    setResult(Constants.RESULT_CAPTURED_PHOTO, intent)
                    finish()
                }
            }
        }
    }

    private fun CameraViewModel.initViewModel() {
        captureNow = { camera.capturePicture() }
        updateGalleryButton = {
            GalleryFragment.loadImages("YogiDiary").run {
                Glide.with(this@CameraActivity)
                        .load(if (isEmpty()) null else this[0].pathOfImage)
                        .error(if (cameraViewModel.isCaptureSizeFull()) {
                            R.drawable.baseline_collections_white_24
                        } else {
                            R.drawable.baseline_collections_black_24
                        })
                        .apply(RequestOptions.circleCropTransform())
                        .diskCacheStrategy(DiskCacheStrategy.NONE)
                        .skipMemoryCache(true)
                        .into(btn_go_gallery)
            }
        }
    }

    private fun startBlinkAnimation() {
        val blinkAnimation = AnimationUtils.loadAnimation(this@CameraActivity, R.anim.blink)
        blinkAnimation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                // Do nothing
            }

            override fun onAnimationEnd(animation: Animation) {
                frame_dark_screen.visibility = View.GONE
            }

            override fun onAnimationRepeat(animation: Animation) {
                // Do nothing
            }
        })

        frame_dark_screen.run {
            visibility = View.VISIBLE
            startAnimation(blinkAnimation)
        }
    }

    private fun setCaptureSize(width: Int, height: Int, x: Int, y: Int) {
        camera.run {
            stop()
            layoutParams.width = width
            layoutParams.height = height
            val ratio = SizeSelectors.aspectRatio(AspectRatio.of(x, y), 0f)
            val result = SizeSelectors.or(ratio, SizeSelectors.biggest())
            setPictureSize(result)
            start()
        }
    }
}